### PR TITLE
pin numpy and fix tf test for dtype

### PR DESCRIPTION
**Pull request by BenWilson2**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Pin numpy for libraries that have internal incompatibilities with deprecated numpy base types that were removed in numpy 1.24.0. 
Adjust a TF test that uses a deprecated dtype classification for the creation of a numpy array

## How is this patch tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [X] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/models/evaluation/base.py [hunk 1]
```diff
@@ -420,15 +420,23 @@ def __init__(self, data, *, targets, name=None, path=None, feature_names=None):
...
+                "If the `data` argument is a numpy array, it must be a 2 dimension "
+                "array and second dimension represent the number of features. If the `data` "
+                "argument is a list, each of its element must be a feature array of "
+                "numpy array or list and all element must has the same length."
+            )
+
             if isinstance(data, list):
-                data = np.array(data)
+                try:
+                    data = np.array(data)
```
Comment (BenWilson2) [inline mlflow/models/evaluation/base.py:433]: numpy 1.24.x no longer allows creating arrays from mixed length lists and will throw a ValueError if supplied invalid structures.

**Review by dbczumar — COMMENTED** [review]
@BenWilson2 instead of pinning numpy in our tests, can we fix all of the incompatibilities that make them fail? Users will run into these incompatibilities when using these features with the latest numpy versions, right?

**Comment by BenWilson2** [comment]
> @BenWilson2 instead of pinning numpy in our tests, can we fix all of the incompatibilities that make them fail? Users will run into these incompatibilities when using these features with the latest numpy versions, right?

The gluon and shap issues are internal failures to those packages that are still using the deprecated numpy data types within their logic. The pins for onnx and mleap are for older versions only; the latest releases of them have fixes in place for numpy's 1.24.x changes. 
The two test failures that we have control over are for the numpy array creation from mixed length lists and the tensorflow array equivalency assertion that are adjusted in this PR.

**Review by WeichenXu123 — APPROVED** [review]

File: mlflow/models/evaluation/base.py [hunk 2]
```diff
@@ -420,15 +420,23 @@ def __init__(self, data, *, targets, name=None, path=None, feature_names=None):
...
+                except ValueError:
+                    raise MlflowException(message=shape_message, error_code=INVALID_PARAMETER_VALUE)
```
Comment (dbczumar) [inline mlflow/models/evaluation/base.py:435]:
```suggestion
                except ValueError as e:
                    raise MlflowException(message=shape_message, error_code=INVALID_PARAMETER_VALUE) from e
```

File: mlflow/ml-package-versions.yml [hunk 3]
```diff
@@ -190,6 +190,9 @@ gluon:
   models:
     minimum: "1.5.1"
     maximum: "1.9.1"
+    requirements:
+      # numpy type declarations are fully deprecated
+      ">= 1.6.0": ["numpy<1.24.0"]
```
Comment (dbczumar) [inline mlflow/ml-package-versions.yml:195]: Can we add upper bounds in all places where we're pinning numpy?

Comment (BenWilson2) [reply mlflow/ml-package-versions.yml]: Inverted the two unbounded tests

File: .github/workflows/master.yml [hunk 4]
```diff
@@ -248,7 +248,7 @@ jobs:
       - name: Install dependencies
         run: |
           source ./dev/install-common-deps.sh
-          pip install pyspark
+          pip install pyspark "numpy<1.24.0"
```
Comment (dbczumar) [inline .github/workflows/master.yml:251]: Can we avoid the pin here? I worry we'll forget to remove it. If these `models` tests are duplicating tests from cross version tests, we should deduplicate by removing them from `models` tests

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/8381/files#r1185639016 is addressed. Thanks @BenWilson2 ! :D

File: .github/workflows/master.yml (hunk 4 above)
Comment (BenWilson2) [reply .github/workflows/master.yml:251]: It's actually for mlflow evaluate and it's use of shap. I really like the idea of splitting that out into its own CI suite, though. We'd have to leave the pin in, but it wouldn't hamper our ability for the much larger scope of mlflow/models to be pinned to an old version of numpy. :) I'm going to push that change now.

File: .github/workflows/master.yml [hunk 5]
```diff
@@ -254,6 +254,28 @@ jobs:
...
+          submodules: recursive
+          repository: ${{ github.event.inputs.repository }}
+          ref: ${{ github.event.inputs.ref }}
+      - uses: ./.github/actions/setup-python
+      - uses: ./.github/actions/setup-pyenv
+      - uses: ./.github/actions/setup-java
+      - name: Install dependencies
+        run: |
+          source ./dev/install-common-deps.sh
+          pip install pyspark "numpy<1.24.0"
```
Comment (dbczumar) [inline .github/workflows/master.yml:272]: Can we leave a comment explaining why we have this pin and file an internal ticket for next sprint to track removing the pin? Feel free to merge after that. Thanks @BenWilson2 !

Comment (BenWilson2) [reply .github/workflows/master.yml:272]: Ticket's in :D (earlier today, just in case ;) ) Adding comment now

**Pull request merged**